crabs/detection_tracking/train_model.py

```diff
@@ -116,7 +118,67 @@ def setup_trainer(self):
             limit_train_batches=self.limit_train_batches,
         )
 
-    def train_model(self):
+    def optuna_maximise_val_precision_recall(
+        self, trial: optuna.Trial
+    ) -> float:
+        """Objective function for Optuna.
+
+        When used with Optuna, it wil maximise precision and recall on the
+        validation set.
+
+        Parameters
+        ----------
+        trial : optuna.Trial
+            The trial to optimise.
+
+        Returns
+        -------
+        float
+            The value to maximise.
+        """
+        # Sample hyperparameters from the search space for this trial
+        # learning rate (a float)
+        if "learning_rate" in self.config["optuna"]:
+            if len(self.config["optuna"]["learning_rate"]) != 2:
+                raise ValueError(
+                    "bounds for `learning_rate` not defined in optuna config"
+                )
+            else:
+                self.config["learning_rate"] = trial.suggest_float(
+                    "learning_rate",
+                    float(self.config["optuna"]["learning_rate"][0]),
+                    float(self.config["optuna"]["learning_rate"][1]),
+                )
+
+        # num_epochs (an integer)
+        if "num_epochs" in self.config["optuna"]:
```

Comment:
this check that both bounds are defined feels like it belongs somewhere else....

we can also remove it 🤔